packages/flutter_tools/lib/src/build_system/source.dart

```diff
@@ -222,12 +226,26 @@ class SourceVisitor implements ResolvedFiles {
     }
     sources.add(entity as File);
   }
+
+  void visitProjectSource(ProjectSourceBuilder builder, bool optional) {
+    final File source = builder(_project);
```

Comment:
In the future, we could allow the builder to return `File`s or `Directory`s.by changing its return type to `FileSystemEntity`. We don't have any scenarios that need this today, so I didn't implement this.